### PR TITLE
feat: add and use a user operation trait

### PR DESCRIPTION
**Pull request by dancoombs**
Introduce a user operation trait. WIP to test the strategy, needs tons of work to actually get ready to merge, however spec tests pass.

- Implement for v0.6 and v0.7
- Add a variant that can hold either, and also implements the trait
- Make UO specific components generic on the trait:
  - Pool classes
  - Bundle proposer and bundle sender
  - Likely will need some RPC helper functions here too.
- Non-UO specific components use the variant 
  -  Local/remote client and server classes
  - Makes it easy to communicate between rpc, pool, and builder actors

Related TODOs:
-  RPC is hardcoded on v0.6. Introduce new RPC types that can distinguish the required UO type based on the entry point address from ChainSpec. Then create the variant and send to the pool OR use the entry point context (only estimator).
- Local pool runner is hardcoded on v0.6. Needs to take the variant and the ep address, then route to the correct mempool while extracting the correct type.
- Entry point provider, simulation, estimation, need 0.7 specific implementations
- Bundle proposer needs to implement 0.7 specific logic (packing UO before sending).

### CODE REVIEW

File: crates/builder/src/bundle_proposer.rs [hunk 1]
```diff
@@ -73,37 +85,41 @@ impl Bundle {
...
     }
 
-    pub(crate) fn iter_ops(&self) -> impl Iterator<Item = &UserOperation> + '_ {
+    pub(crate) fn iter_ops(&self) -> impl Iterator<Item = &UO> + '_ {
         self.ops_per_aggregator.iter().flat_map(|ops| &ops.user_ops)
     }
 }
 
-#[cfg_attr(test, automock)]
+#[cfg_attr(test, automock(type UO = rundler_types::UserOperationVariant;))]
```
Comment (dancoombs) [inline crates/builder/src/bundle_proposer.rs:93]: TODO: look into these, is this the correct way to handle these automocks?

File: crates/builder/src/bundle_proposer.rs [hunk 2]
```diff
@@ -73,37 +85,41 @@ impl Bundle {
...
+    S: Simulator<UO = UO>,
+    E: EntryPoint<UO = UO>,
     P: Provider,
     C: PoolServer,
 {
     builder_index: u64,
     pool: C,
     simulator: S,
-    entry_point: E,
+    entry_point: Arc<E>,
```
Comment (dancoombs) [inline crates/builder/src/bundle_proposer.rs:117]: TODO: Should this need to be `Arc`

File: crates/builder/src/bundle_proposer.rs [hunk 3]
```diff
@@ -452,12 +467,8 @@ where
             }
 
             // Update the running gas that would need to be be spent to execute the bundle so far.
-            gas_spent += gas::user_operation_execution_gas_limit(
-                &self.settings.chain_spec,
-                &op,
-                false,
-                simulation.requires_post_op,
-            );
+            gas_spent +=
```
Comment (dancoombs) [inline crates/builder/src/bundle_proposer.rs:470]: TODO: removed `simulation.requires_post_op` from this function, ensure this still works.

File: crates/builder/src/bundle_proposer.rs [hunk 4]
```diff
@@ -1039,31 +1065,20 @@ impl ProposalContext {
...
-            );
+            let required_gas = get_gas_required_for_op(chain_spec, gas_spent, op);
             max_gas = cmp::max(max_gas, required_gas);
-            gas_spent += gas::user_operation_gas_limit(
-                chain_spec,
-                op,
-                false,
-                op_with_sim.simulation.requires_post_op,
-            );
+            gas_spent += gas::user_operation_gas_limit(chain_spec, op, false);
```
Comment (dancoombs) [inline crates/builder/src/bundle_proposer.rs:1070]: TODO: Same here, removed post op

File: crates/builder/src/bundle_proposer.rs [hunk 5]
```diff
@@ -1226,24 +1241,16 @@ impl ProposalContext {
...
-
     gas_spent
         + gas::user_operation_pre_verification_gas_limit(chain_spec, op, false)
-        + op.verification_gas_limit * 2
-        + op.call_gas_limit
-        + post_exec_req_gas
+        + op.total_verification_gas_limit()
+        + op.required_pre_execution_buffer()
+        + op.call_gas_limit()
 }
```
Comment (dancoombs) [inline crates/builder/src/bundle_proposer.rs:1254]: In general: we should split this file up

File: crates/builder/src/task.rs [hunk 6]
```diff
@@ -23,9 +23,9 @@ use ethers_signers::Signer;
 use futures::future;
 use futures_util::TryFutureExt;
 use rundler_pool::PoolServer;
-use rundler_provider::{EntryPoint, EthersEntryPoint};
+use rundler_provider::EthersEntryPointV0_6;
 use rundler_sim::{
-    MempoolConfig, PriorityFeeMode, SimulateValidationTracerImpl, SimulationSettings, SimulatorImpl,
+    simulation::v0_6 as sim_v0_6, MempoolConfig, PriorityFeeMode, SimulationSettings,
```
Comment (dancoombs) [inline crates/builder/src/task.rs:28]: Todo: Just call this `SimulatorV0_6` in this file

File: crates/builder/src/task.rs [hunk 7]
```diff
@@ -23,9 +23,9 @@ use ethers_signers::Signer;
 use futures::future;
 use futures_util::TryFutureExt;
 use rundler_pool::PoolServer;
-use rundler_provider::{EntryPoint, EthersEntryPoint};
+use rundler_provider::EthersEntryPointV0_6;
```
Comment (dancoombs) [inline crates/builder/src/task.rs:26]: TODO: move this to `v0_6::EthersEntryPoint` then do `v0_6::EthersEntryPoint as EthersEntryPointV0_6 `

File: crates/pool/src/mempool/error.rs [hunk 8]
```diff
@@ -108,7 +108,7 @@ impl From<PrecheckError> for MempoolError {
         // extract violation and replace with dummy
         Self::PrecheckViolation(mem::replace(
             violation,
-            PrecheckViolation::InitCodeTooShort(0),
```
Comment (dancoombs) [inline crates/pool/src/mempool/error.rs:111]: TODO: check for this in RPC module

File: crates/pool/src/mempool/mod.rs [hunk 9]
```diff
@@ -291,6 +291,51 @@ impl PoolOperation {
     }
 }
 
+// TODO had to hardcode these because trying to use the generic conflicts with the core `impl<T> From<T> for T` impl
```
Comment (dancoombs) [inline crates/pool/src/mempool/mod.rs:294]: TODO: remove

File: crates/pool/src/mempool/mod.rs [hunk 10]
```diff
@@ -291,6 +291,51 @@ impl PoolOperation {
...
+            sim_block_number: op.sim_block_number,
+            entities_needing_stake: op.entities_needing_stake,
+            account_is_staked: op.account_is_staked,
+            entity_infos: op.entity_infos,
+        }
+    }
+}
+
+impl From<PoolOperation<v0_6::UserOperation>> for PoolOperation<UserOperationVariant> {
+    fn from(op: PoolOperation<v0_6::UserOperation>) -> PoolOperation<UserOperationVariant> {
```
Comment (dancoombs) [inline crates/pool/src/mempool/mod.rs:323]: TODO: can we make this generic somehow?

File: crates/provider/src/ethers/entry_point_v0_6.rs [hunk 11]
```diff
@@ -240,14 +265,110 @@ where
...
+            .nonce(U256::from(100_000))
+            .chain_id(U64::from(100_000))
+            .rlp();
+
+        let l1_fee = self.opt_gas_oracle.get_l1_fee(tx).call().await?;
+        Ok(l1_fee.checked_div(gas_price).unwrap_or(U256::MAX))
+    }
 }
 
+// TODO: pack for 0.7
```
Comment (dancoombs) [inline crates/provider/src/ethers/entry_point_v0_6.rs:357]: Remove

File: crates/provider/src/ethers/mod.rs [hunk 12]
```diff
@@ -13,7 +13,7 @@
 
 //! Provider implementations using [ethers-rs](https://github.com/gakonst/ethers-rs)
 
-mod entry_point;
-pub use entry_point::EntryPointImpl as EthersEntryPoint;
+mod entry_point_v0_6;
+pub use entry_point_v0_6::EntryPointV0_6 as EthersEntryPointV0_6;
```
Comment (dancoombs) [inline crates/provider/src/ethers/mod.rs:17]: change to `entrypoint::v0_6`

File: crates/rpc/src/eth/api.rs [hunk 13]
```diff
@@ -102,13 +102,13 @@ pub(crate) struct EthApi<P, E, PS> {
 impl<P, E, PS> EthApi<P, E, PS>
 where
     P: Provider,
-    E: EntryPoint,
+    E: EntryPoint<UO = v0_6::UserOperation>,
```
Comment (dancoombs) [inline crates/rpc/src/eth/api.rs:105]: Follow up PRs will remove this dependance and route to the correct entry point based on address

File: crates/rpc/src/task.rs [hunk 14]
```diff
@@ -147,20 +148,24 @@ where
...
+    {
         for api in &self.args.api_namespaces {
             match api {
                 ApiNamespace::Eth => module.merge(
                     EthApi::new(
                         self.args.chain_spec.clone(),
                         provider.clone(),
                         // TODO: support multiple entry points
-                        vec![entry_point.clone()],
+                        vec![Arc::new(entry_point.clone())],
```
Comment (dancoombs) [inline crates/rpc/src/task.rs:168]: TODO: figure out this `Arc`

Comment (dancoombs) [reply crates/rpc/src/task.rs:168]: (shouldn't need arc if it implements clone)

File: crates/rpc/src/types.rs [hunk 15]
```diff
@@ -259,3 +259,25 @@ pub struct RpcDebugPaymasterBalance {
     /// Paymaster confirmed balance onchain
     pub confirmed_balance: U256,
 }
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct RpcGasEstimate {
```
Comment (dancoombs) [inline crates/rpc/src/types.rs:265]: TODO: pub(crate) in this file

File: crates/sim/src/estimation/v0_6/estimator.rs [hunk 16]
```diff
@@ -38,13 +36,16 @@ use rundler_types::{
...
+    GasEstimate,
 };
 use rundler_utils::{eth, math};
 use tokio::join;
 
-use super::types::{GasEstimate, Settings, UserOperationOptionalGas};
-use crate::{gas, precheck::MIN_CALL_GAS_LIMIT, simulation, utils, FeeEstimator};
+use crate::{
+    estimation::GasEstimationError, gas, precheck::MIN_CALL_GAS_LIMIT, simulation, utils,
+    EstimationSettings as Settings, FeeEstimator,
```
Comment (dancoombs) [inline crates/sim/src/estimation/v0_6/estimator.rs:47]: Fix this import

File: crates/sim/src/estimation/v0_6/mod.rs [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+//
+// Rundler is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along with Rundler.
+// If not, see https://www.gnu.org/licenses/.
+
+mod estimator;
+pub use estimator::GasEstimator;
```
Comment (dancoombs) [inline crates/sim/src/estimation/v0_6/mod.rs:15]: Is this file needed? Can we just have the estimator?

File: crates/sim/src/gas/gas.rs [hunk 18]
```diff
@@ -149,102 +123,58 @@ pub async fn calc_required_pre_verification_gas<P: Provider>(
 /// If limiting the size of a bundle transaction to adhere to block gas limit, use the execution gas limit functions.
 
 /// Returns the gas limit for the user operation that applies to bundle transaction's limit
-pub fn user_operation_gas_limit(
+pub fn user_operation_gas_limit<UO: UserOperation>(
```
Comment (dancoombs) [inline crates/sim/src/gas/gas.rs:126]: Check these changes for correctness

File: crates/sim/src/gas/gas.rs [hunk 19]
```diff
@@ -27,32 +27,6 @@ use super::oracle::{
     ConstantOracle, FeeOracle, ProviderOracle, UsageBasedFeeOracle, UsageBasedFeeOracleConfig,
 };
 
-/// Gas overheads for user operations used in calculating the pre-verification gas. See: https://github.com/eth-infinitism/bundler/blob/main/packages/sdk/src/calcPreVerificationGas.ts
```
Comment (dancoombs) [inline crates/sim/src/gas/gas.rs:30]: Need to move this to ChainSpec

File: crates/sim/src/simulation/mod.rs [hunk 20]
```diff
@@ -11,18 +11,474 @@
...
+    OutOfGas(Entity),
+    /// The user operation aggregator signature validation failed
+    #[display("aggregator signature validation failed")]
+    AggregatorValidationFailed,
+    /// Verification gas limit doesn't have the required buffer on the measured gas
+    #[display("verification gas limit doesn't have the required buffer on the measured gas, limit: {0}, needed: {1}")]
+    VerificationGasLimitBufferTooLow(U256, U256),
+}
+
+/// A wrapper around Opcode that implements extra traits
```
Comment (dancoombs) [inline crates/sim/src/simulation/mod.rs:184]: TODO: move a lot of this to the types file

Comment (dancoombs) [reply crates/sim/src/simulation/mod.rs:184]: OR move type file here

File: crates/types/src/user_operation/mod.rs [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+pub struct UserOpsPerAggregator<UO: UserOperation> {
+    /// User operations
+    pub user_ops: Vec<UO>,
+    /// Aggregator address, zero if no aggregator is used
+    pub aggregator: Address,
+    /// Aggregator signature, empty if no aggregator is used
+    pub signature: Bytes,
+}
+
+// TODO(danc): move this to chain spec
```
Comment (dancoombs) [inline crates/types/src/user_operation/mod.rs:329]: TODO

File: crates/types/src/user_operation/v0_6.rs [hunk 22]
```diff
@@ -0,0 +1,444 @@
...
+    pub max_fee_per_gas: Option<U256>,
+    /// Max priority fee per gas (optional, ignored if set)
+    pub max_priority_fee_per_gas: Option<U256>,
+    /// Paymaster and data (required, dummy value for gas estimation)
+    pub paymaster_and_data: Bytes,
+    /// Signature (required, dummy value for gas estimation)
+    pub signature: Bytes,
+}
+
+impl UserOperationOptionalGas {
```
Comment (dancoombs) [inline crates/types/src/user_operation/v0_6.rs:225]: A future gas estimation PR may turn this into an interface as well. We also may just want to move these functions into the gas estimation crate as a wrapper impl as they are only used there.

File: crates/types/src/user_operation/v0_7.rs [hunk 23]
```diff
@@ -0,0 +1,412 @@
...
+            ret.push(Entity::factory(factory));
+        }
+        if let Some(paymaster) = self.paymaster {
+            ret.push(Entity::paymaster(paymaster));
+        }
+        ret
+    }
+
+    fn heap_size(&self) -> usize {
+        unimplemented!()
```
Comment (dancoombs) [inline crates/types/src/user_operation/v0_7.rs:120]: TODO lots of unimplemented in this file

File: crates/types/src/user_operation/v0_7.rs [hunk 24]
```diff
@@ -0,0 +1,412 @@
...
+        };
+
+        let packed = pack_user_operation(uo.clone());
+        let hash = hash_packed_user_operation(packed.clone(), self.entry_point, self.chain_id);
+
+        UserOperation { hash, packed, ..uo }
+    }
+}
+
+fn pack_user_operation(uo: UserOperation) -> PackedUserOperation {
```
Comment (dancoombs) [inline crates/types/src/user_operation/v0_7.rs:324]: TODO: unpack

File: crates/sim/src/estimation/v0_6/mod.rs (hunk 17 above)
Comment (dancoombs) [reply crates/sim/src/estimation/v0_6/mod.rs:15]: Leaving it

File: crates/builder/src/bundle_proposer.rs [hunk 25]
```diff
@@ -1034,36 +1061,27 @@ impl ProposalContext {
...
-            max_gas = cmp::max(max_gas, required_gas);
-            gas_spent += gas::user_operation_gas_limit(
-                chain_spec,
-                op,
-                false,
-                op_with_sim.simulation.requires_post_op,
-            );
-        }
-        max_gas
+        // TODO(danc): in the 0.7 entrypoint we could optimize this by removing the need for
```
Comment (dancoombs) [inline crates/builder/src/bundle_proposer.rs:1064]: Add a ticket to follow up here when we do bundle building improvements. For now lets just ensure this is correct, if not optimal.

Comment (dancoombs) [reply crates/builder/src/bundle_proposer.rs:1064]: https://github.com/alchemyplatform/rundler/issues/639

File: crates/rpc/src/eth/mod.rs [hunk 26]
```diff
@@ -41,10 +41,10 @@ pub trait EthApi {
     #[method(name = "estimateUserOperationGas")]
     async fn estimate_user_operation_gas(
         &self,
-        op: UserOperationOptionalGas,
+        op: v0_6::UserOperationOptionalGas,
```
Comment (0xfourzerofour) [inline crates/rpc/src/eth/mod.rs:44]: I though the RPC layer would be some type of abstraction to both 0.6 and 0.7?

File: crates/sim/src/estimation/v0_6.rs [hunk 27]
```diff
@@ -110,7 +83,13 @@ pub struct GasEstimatorImpl<P, E> {
 }
 
 #[async_trait::async_trait]
-impl<P: Provider, E: EntryPoint> GasEstimator for GasEstimatorImpl<P, E> {
+impl<P, E> crate::estimation::GasEstimator for GasEstimator<P, E>
```
Comment (0xfourzerofour) [inline crates/sim/src/estimation/v0_6.rs:86]: Same here, not a big fan of using the full import path in the trait implementation

File: crates/sim/src/precheck.rs [hunk 28]
```diff
@@ -155,47 +170,37 @@ impl<P: Provider, E: EntryPoint> PrecheckerImpl<P, E> {
             settings,
             fee_estimator,
             cache: RwLock::new(AsyncDataCache { fees: None }),
+            _uo_type: std::marker::PhantomData,
```
Comment (0xfourzerofour) [inline crates/sim/src/precheck.rs:173]: Same here with the import path

File: crates/types/src/user_operation/mod.rs [hunk 29]
```diff
@@ -0,0 +1,371 @@
...
+
+    /// Clear the signature field of the user op
+    ///
+    /// Used when a user op is using a signature aggregator prior to being submitted
+    fn clear_signature(&mut self);
+}
+
+/// User operation type enum
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum UserOperationType {
```
Comment (0xfourzerofour) [inline crates/types/src/user_operation/mod.rs:116]: There are two enums UserOperationType and EntryPointVersion. Do you think there is a way we could combine them? they sorta achieve the same thing just talking about seperate parts Entrypoint vs UO

File: crates/types/src/user_operation/v0_7.rs [hunk 30]
```diff
@@ -0,0 +1,613 @@
...
+
+    fn paymaster(&self) -> Option<Address> {
+        self.paymaster
+    }
+
+    fn factory(&self) -> Option<Address> {
+        self.factory
+    }
+
+    fn max_gas_cost(&self) -> U256 {
```
Comment (0xfourzerofour) [inline crates/types/src/user_operation/v0_7.rs:105]: Does 0.7 not have the 3x variable paymaster calculation?

File: crates/types/src/user_operation/v0_7.rs [hunk 31]
```diff
@@ -0,0 +1,613 @@
...
+
+impl From<UserOperation> for super::UserOperationVariant {
+    fn from(op: UserOperation) -> Self {
+        super::UserOperationVariant::V0_7(op)
+    }
+}
+
+/// User Operation with optional gas for Entry Point v0.7
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct UserOperationOptionalGas {
```
Comment (0xfourzerofour) [inline crates/types/src/user_operation/v0_7.rs:222]: Is this just for estimation?

**Review by 0xfourzerofour — COMMENTED** [review]
Pretty wild PR good job on it. I wouldn't really change much just a couple nit things and questions

File: crates/pool/src/mempool/mod.rs [hunk 32]
```diff
@@ -291,9 +291,62 @@ impl PoolOperation {
     }
 }
 
+/// Trait to convert a [PoolOperation] holding a [UserOperationVariant] to a [PoolOperation] with a different user operation type.
+pub trait FromPoolOperationVariant {
```
[the remark this answers is not in the document]
Comment (0xfourzerofour) [reply crates/pool/src/mempool/mod.rs:295]: Actually yeh this is fine because of the variant within the struct. Forget about it

[the remark this answers is not in the document]
Comment (dancoombs) [reply crates/pool/src/mempool/mod.rs:295]: Ended up removing this in a follow up PR. PoolOperation now only holds the variant

File: crates/rpc/src/eth/mod.rs (hunk 26 above)
Comment (dancoombs) [reply crates/rpc/src/eth/mod.rs:44]: Implemented this in https://github.com/alchemyplatform/rundler/pull/636.

This PR mostly implemented the types but hardcoded v0.6 in a bunch of places. Follow up PRs do "routing" based on entry point to the correct implementations.

File: crates/types/src/user_operation/mod.rs (hunk 29 above)
Comment (dancoombs) [reply crates/types/src/user_operation/mod.rs:116]: Yeah I can probably remove this one and just use `EntryPointVersion`

File: crates/types/src/user_operation/v0_7.rs (hunk 30 above)
Comment (dancoombs) [reply crates/types/src/user_operation/v0_7.rs:105]: Nope! Big upgrade. Since they split out paymaster validation gas and paymaster post op gas, you can just total the fields.

File: crates/types/src/user_operation/v0_7.rs (hunk 31 above)
Comment (dancoombs) [reply crates/types/src/user_operation/v0_7.rs:222]: Yeah thats currently the only place where its used

File: crates/provider/src/ethers/entry_point/v0_6.rs [hunk 33]
```diff
@@ -20,105 +20,185 @@ use ethers::{
...
+                Arc::clone(&provider),
+            ),
+            opt_gas_oracle: GasPriceOracle::new(OPTIMISM_BEDROCK_GAS_ORACLE_ADDRESS, provider),
         }
     }
 }
 
 #[async_trait::async_trait]
-impl<P> EntryPoint for EntryPointImpl<P>
+impl<P> crate::traits::EntryPoint for EntryPoint<P>
```
[the remark this answers is not in the document]
Comment (dancoombs) [reply crates/provider/src/ethers/entry_point/v0_6.rs:104]: Fixed

File: crates/sim/src/estimation/v0_6.rs (hunk 27 above)
Comment (dancoombs) [reply crates/sim/src/estimation/v0_6.rs:82]: Fixed

File: crates/sim/src/precheck.rs (hunk 28 above)
Comment (dancoombs) [reply crates/sim/src/precheck.rs:195]: Fixed

File: crates/types/src/user_operation/mod.rs (hunk 29 above)
Comment (dancoombs) [reply crates/types/src/user_operation/mod.rs:135]: Fixed
